sb/robot/sim_robot.py

```diff
@@ -36,21 +41,157 @@ def power(self, value):
+        except KeyError:
+            print("There's no ultrasound module on those pins. Try:")
+            for (
+                    (trigger_pin, echo_pin),
+                    (direction, _),
+            ) in self.ULTRASOUND_ANGLES.items():
+                print("Pins {} and {} for the sensor pointing {}".format(
+                    trigger_pin,
+                    echo_pin,
+                    direction,
+                ))
+            return 0.0
 
+        result = self.robot.send_ultrasound_ping(angle_offset)
+
+        if result is None:
+            # No detection is equivalent to just not getting an echo response
+            result = 0.0
+
+        return result
+
+
+class Camera:
+    def __init__(self, robot):
+        self.robot = robot
+
+    def __str__(self):
+        return "Camera"
+
+    def see(self):
+        with self.robot.lock:
+            x, y = self.robot.location
+            heading = self.robot.heading
+
+        acq_time = time.time()
+        # Block for a realistic amount of time
+        time.sleep(0.2)
+
+        MOTION_BLUR_SPEED_THRESHOLD = 5
+
+        def robot_moving(robot):
+            vx, vy = robot._body.linear_velocity
+            return hypot(vx, vy) > MOTION_BLUR_SPEED_THRESHOLD
+
+        def motion_blurred(o):
+            # Simple approximation: we can't see anything if either it's moving
+            # or we're moving. This doesn't handle tokens grabbed by other robots
+            # but Sod's Law says we're likely to see those anyway.
+            return (robot_moving(self.robot) or
+                    isinstance(o, SimRobot) and robot_moving(o))
+
+        def object_filter(o):
+            # Choose only marked objects within the field of view
+            direction = atan2(o.location[1] - y, o.location[0] - x)
+            return (o.marker_id is not None and
+                    o is not self and
+                    -HALF_FOV_WIDTH < direction - heading < HALF_FOV_WIDTH and
+                    not motion_blurred(o))
+
+        def is_wall_marker(marker_id):
+            return marker_id in WALL
+
+        def is_token_marker(marker_id):
+            return marker_id in TOKEN
+
+        def marker_map(o):
+            # Turn a marked object into a Marker
+            rel_x, rel_y = (o.location[0] - x, o.location[1] - y)
+            rot_y = atan2(rel_y, rel_x) - heading
+            polar_coord = PolarCoord(
+                distance_meters=hypot(rel_x, rel_y),
+                rot_y_rad=rot_y,
+                rot_y_deg=degrees(rot_y)
+            )
+
+            # TODO: Check polar coordinates are the right way around
+            mid = o.marker_id
+            return Marker(
+                id=mid,
+                size=MARKER_SIZES[mid],
+                is_wall_marker=lambda: is_wall_marker(mid),
+                is_token_marker=lambda: is_token_marker(mid),
+                polar=polar_coord,
+                pixel_centre="NOT SUPPORTED",
```

Comment:
I think it would be better to just not have these functions at all rather than this hacky string